### ISSUE TITLE
Neighbourhood grouping

### ISSUE THREAD

**Issue by MikeDMorgan** [MarioniLab/miloR#45]
Except in the case of the multiple discrete blocks, the neighbourhood grouping _almost always_ only generates 2 groups if I use the same FDR threshold as the DA testing (relaxing this gives more granular grouping in line with expectations in the mouse thymus data). Even an FDR 100% still only returns 4 groups...

My concern is that the simple approach using weakly connected components struggles when the graph is highly connected. I propose that we use a community detection e.g. Louvain/Infomap/Walktrap or something to get more granularity in the grouping perhaps?

**Comment by MikeDMorgan** [comment]
I should say, that this is obviously alleviated by subsetting the neighbourhoods first - but users aren't always going to have a grouping _a priori_ like we do for the public data sets...

**Comment by emdann** [comment]
But isn't the number of groups mainly determined by the amount of overlap?

**Comment by MikeDMorgan** [comment]
It doesn't seem so. I think in most continuous trajectories, or where there is quite high similarity in transcriptomes (as is the case for TEC, T cells and others), then they are not that weakly connected.

**Comment by MikeDMorgan** [comment]
Changing the overlap from 1 to 150 still gives the same groups!!

**Comment by MikeDMorgan** [comment]
Maybe use a hierarchical clustering on the neighbourhood similarity matrix.

**Comment by MikeDMorgan** [comment]
I've done a little big more digging using the simulated trajectories and mouse thymus data sets. I've run the neighbourhood grouping code using a range of overlapping cell thresholds. It seems that the graph component based grouping is _very_ sensitive to the specific threshold used.

Below I'm showing all neighbourhoods in the simulated trajectory data, where the x-axis is the `overlap` threshold and the y-axis is the total number of groups returned.  These data consist of 500 simulated cells.

![Simulation_GroupingThreshold_curve](https://user-images.githubusercontent.com/6473398/97027728-11cc4280-1553-11eb-8436-d1ebd874a236.png)

The steep inflection indicates that it doesn't take much to break up this graph. I've also plotted the same curve but just for the number of groups returned that contain DA neighbourhoods.

![Simulation_GroupingThreshold_curve-Sig](https://user-images.githubusercontent.com/6473398/97027862-3d4f2d00-1553-11eb-92f0-2a6a890b4ff1.png)

The shapes of these curves are very similar - this data set should only have 1 DA neighbourhood group.

As a comparison I ran the same analysis on the mouse thymus data.

![Thymus_GroupingThreshold_curve](https://user-images.githubusercontent.com/6473398/97027990-653e9080-1553-11eb-9538-25b8dcc9284d.png)

These data are _very_ sensitive to the `overlap` value used: the difference between `overlap=10` (7 groups) and `overlap=`15` (27 groups) is quite striking. Below is the equivalent plot for the sig DA neighbourhood groups.

![Thymus_GroupingThreshold_curve-Sig](https://user-images.githubusercontent.com/6473398/97028210-b8184800-1553-11eb-9e4a-4b3494ee8859.png)

This has me a little concerned, as varying when `overlap` is set in the function, it doesn't actually return the expected number of groups. e.g. using the discrete simulated data set.

```
> table(da.res$SpatialFDR < 0.1)

FALSE  TRUE 
   21    20 
> discrete.markers <- findNhoodMarkers(milo, da.res=da.res, da.fdr=0.1, overlap=30)
Found 20 DA neighbourhoods at FDR 10%
nhoodAdjacency found - using for nhood grouping
Nhoods aggregated into 1 groups
```

But when running the code outside of the function I get:
```
> head(t.df[t.df$Thresh %in% c(20:30), ], 20)
   NGroups NSigGroups Thresh
20       4          2     20
21       6          4     21
22       6          4     22
23       6          4     23
24       8          6     24
25       8          6     25
26       9          7     26
27      10          8     27
28      10          8     28
29      11          9     29
30      12         10     30
```

i.e. the function should be returning 2 significant groups, but only returns 1!

I will continue to debug this next week, but its definitely not doing what it should be...
